web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts

```diff
@@ -54,42 +60,219 @@ export class AccessRequestsService {
     });
   }
 
-  getAddedResourceCount() {
+  getAddedResourceCount(): number {
     const pendingAccessRequest = this.getState().pending;
-    return (
-      Object.keys(pendingAccessRequest.node).length +
-      Object.keys(pendingAccessRequest.db).length +
-      Object.keys(pendingAccessRequest.app).length +
-      Object.keys(pendingAccessRequest.kube_cluster).length +
-      Object.keys(pendingAccessRequest.windows_desktop).length +
-      Object.keys(pendingAccessRequest.user_group).length
-    );
+    if (pendingAccessRequest.kind === 'role') {
+      return pendingAccessRequest.roles.size;
+    }
+    if (pendingAccessRequest.kind === 'resource') {
+      return pendingAccessRequest.resources.size;
+    }
+
+    return 0;
+  }
+
+  async addOrRemoveResource({
+    kind,
+    resource,
+  }: ResourceRequest): Promise<void> {
+    this.setState(draftState => {
+      if (draftState.pending.kind !== 'resource') {
+        draftState.pending = {
+          kind: 'resource',
+          resources: new Map(),
+        };
+      }
+
+      const { resources } = draftState.pending;
+
+      if (resources.has(resource.uri)) {
+        resources.delete(resource.uri);
+      } else {
+        // Store only properties required by the type.
+        if (kind === 'app' || kind === 'database' || kind === 'kube') {
```

Comment:
Wouldn't it be better to do it the other way around? As in do `if (kind === 'server')` followed by `else`? Otherwise each new resource type we add will have hostname added.

This too could be a switch but idk if it really needs to be. As long as we have a sensible default that could in theory be applied to each new kind of resource, then I think it's fine to have a simple if else. In this case we do have such a default – each resource must have its own URI.